### PR TITLE
[router] fix initial url on web when using baseUrl

### PR DESCRIPTION
**Pull request by marklawlor**
# Why

ENG-11108

When using a baseUrl on web after the initial load a `history.replace` event is fired which strips the baseUrl. This is caused by React Navigation examining the current URL (which has the baseURL) and the `url` attribute on the current focused route (which does not include the baseURL). As they are different, it assumes the URL is wrong and changes it via `history.replace`.

# How

I tried to fix this within the existing code, but eventually concluded that we simply need to fork `useLInking` for web as well. This fork simply removes the logic to use the URL on the state, and fall back to `getPathFromState` (which correctly handles baseUrl)

# Test Plan

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: packages/expo-router/src/fork/useLinking.ts [hunk 1]
```diff
@@ -1,3 +1,372 @@
-import useLinking from '@react-navigation/native/lib/module/useLinking';
+/* eslint-disable */
+// Forked from react-navigation to add basePath functionality to web.
+// By default, on load React Navigation uses the path stored in the state. That code is commented out
+// so it falls back to calling getPathFromState
+// https://github.com/react-navigation/react-navigation/blob/6.x/packages/native/src/useLinking.tsx
+// Changed:
+//   - https://github.com/react-navigation/react-navigation/blob/13d4aa270b301faf07960b4cd861ffc91e9b2c46/packages/native/src/useLinking.tsx#L264-L283
```
Comment (marklawlor) [inline packages/expo-router/src/fork/useLinking.ts:7]: Need to clarify that this `if` statement was removed

**Comment by marklawlor** [comment]
Readded focused route logic. Just append the baseURL to this line of code https://github.com/react-navigation/react-navigation/blob/13d4aa270b301faf07960b4cd861ffc91e9b2c46/packages/native/src/useLinking.tsx#L280

File: packages/expo-router/src/fork/useLinking.ts [hunk 2]
```diff
@@ -1,3 +1,405 @@
-import useLinking from '@react-navigation/native/lib/module/useLinking';
+/* eslint-disable */
```
Comment (marklawlor) [inline packages/expo-router/src/fork/useLinking.ts:1]: All Forked sections have `Start of fork/End of fork` code comments

File: packages/expo-router/src/fork/useLinking.ts [hunk 3]
```diff
@@ -1,3 +1,405 @@
...
+    getStateFromPathRef.current = getStateFromPath;
+    getPathFromStateRef.current = getPathFromState;
+    getActionFromStateRef.current = getActionFromState;
+  });
+
+  /* Start of fork. Source: https://github.com/react-navigation/react-navigation/blob/13d4aa270b301faf07960b4cd861ffc91e9b2c46/packages/native/src/useLinking.tsx#L142 */
+  // ServerContext is used inside ServerContainer to set the location during SSR: https://github.com/react-navigation/react-navigation/blob/13d4aa270b301faf07960b4cd861ffc91e9b2c46/packages/native/src/ServerContainer.tsx#L50-L54
+  // Expo Router uses the `initialLocation` prop to set the initial location during SSR:
+  const location = React.useContext(ServerLocationContext);
+  const server = { location };
```
Comment (marklawlor) [inline packages/expo-router/src/fork/useLinking.ts:156]: @EvanBacon I'm not sure this behaviour is correct.

`ServerContext` is only used in React Navigation inside the `ServerContainer` to set the initial URL https://github.com/react-navigation/react-navigation/blob/13d4aa270b301faf07960b4cd861ffc91e9b2c46/packages/native/src/ServerContainer.tsx#L50-L54

But Expo Router never sets the location during `renderToStaticContent` 
https://github.com/expo/expo/blob/01a47a80611582494484ee6954d6f5ce6a860131/packages/expo-router/src/static/renderStaticContent.tsx#L108

If I were to replicate this current behaviour exactly, I should just remove `ServerContext`. Instead I've replaced it with a new context (the Provider sets the location in ExpoRoot)

**Comment by marklawlor** [comment]
Note: Test without JavaScript enabled

**Review by EvanBacon — APPROVED** [review]

**Pull request merged**